src/kernels/MIOpenDropoutHIP.cpp

```diff
@@ -0,0 +1,374 @@
+ */
+extern "C" __global__ void
+InitKernelStateHIP(rocrand_state_xorwow* state, ulong prng_seed, ulong states_num)
+{
+    // Get the index of the current element
+    size_t index  = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t stride = blockDim.x * gridDim.x;
+
+    for(size_t gid = index; gid < states_num; gid += stride)
+    {
+        rocrand_state_xorwow state_gid;
+        rocrand_init(prng_seed, gid, 0ULL, &state_gid);
+        state[gid] = state_gid;
+    }
+}
+#endif
+
+#if !RUN_INIT_PRNG
+
+#if RUN_FORWARD
+#ifndef USE_MASK
+#define USE_MASK 0
+#endif
+
+#ifndef USE_RSVSP
+#define USE_RSVSP 0
+#endif
+
+// Dropout Forward
+/**
+ * @brief Data is loaded into registers from the input.
+ * The mask is either generated and saved in the reserved space, or an existing mask is loaded into
+ * registers from the reserved space. The loaded mask is applied to the data in the registers. The
+ * processed data is then written to the output. Reading and writing of input data can be performed
+ * as vectorized operations, with the extent of vectorization determined by `#RD_BLOCK`.
+ */
+template <typename F, typename T, typename B, bool MASK = false, bool RSVSP = false>
+__forceinline__ __device__ void dropoutfw(const rocrand_state_xorwow* state,
+                                          float dropout,
+                                          float scale,
+                                          int dim1,
+                                          int dim2,
+                                          int dim3,
+                                          int dim4,
+                                          F* y,
+                                          int out_str0,
+                                          int out_str1,
+                                          int out_str2,
+                                          int out_str3,
+                                          const F* x,
+                                          int in_str0,
+                                          int in_str1,
+                                          int in_str2,
+                                          int in_str3,
+                                          uchar* reserveSpace,
+                                          size_t total_work,
+                                          size_t in_offset,
+                                          size_t out_offset,
+                                          size_t rsvsp_offset)
+{
+    F dat_blk[RD_BLCK];     // Register space to read the input data
+    uchar is_kept[RD_BLCK]; // Register space to store the mask for the dropout
+
+    size_t sid = threadIdx.x + blockIdx.x * blockDim.x;
+    rocrand_state_xorwow cur_state; // Read the state of the current thread
+    cur_state = state[sid];
+
+    for(size_t gid = threadIdx.x + blockIdx.x * blockDim.x; gid < total_work;
+        gid += blockDim.x * gridDim.x)
+    {
+        size_t i0    = gid / (dim1 * dim2 * dim3 * dim4);
```

Comment:
I've seen this type of computation (converting `gid` to a tensor index) being done a few times. We should make a note for now and create a helper function or a class for this kind of conversion.

Reply:
@amberhassaan 
As far as I can see for this particular kernel, it is used twice - in forward and backward kernels. And we are going to get rid of separated backward kernel.

As for the overall code duplication across the algorithms, I'm afraid that all the algorithms have some small tweaks and differences, and indexing approaches, so it can be hard to generalize a helper for it.

I would suggest to create another issue to evaluate possibility for the helper function across all the algorithms and most probably it must be done when we rewrite all the OCL kernels.